README.md
Giggle.js
=========

[![Build Status](https://travis-ci.org/valeriansaliou/giggle.svg?branch=master)](https://travis-ci.org/valeriansaliou/giggle) [![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/valeriansaliou/giggle?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

**Giggle library**, implementation of [XEP-0166 (Jingle)](http://xmpp.org/extensions/xep-0166.html). _XMPP audio/video chat in your Web browser!_

To be used with JSJaC from Stefan Strigler, available there: https://github.com/sstrigler/JSJaC
Giggle is pluggable to other third-party XMPP libraries, please refer to `./src/giggle.plug.js`. We'll happily accept your Pull Requests (PR).

## How To Build It?

Only Giggle.js raw sources are provided in this repository. Those are not usable out-of-the-box for production purposes. You first need to build the library components on your local machine.

**Important: our automated build system requires that you have NodeJS, NPM, GruntJS, Bower (+ bower-installer) and Java on your machine. UNIX systems (MacOS, Linux) are recommended to proceed build.**

If you already have NodeJS and NPM installed, you can simply install GruntJS and Bower:

> npm install -g grunt-cli bower bower-installer

Deploy the library in one simple command:

> ./tools/deploy.sh

## Who's Behind It?

Made with love by the happy folks at **Jappix** https://jappix.com/, originally for the awesome guys at **Uno IM** https://uno.im/

## Testing Giggle

Please first clone Giggle on your local environment, and access it over your local HTTP server. **Serving example pages over HTTP is a requirement for WebRTC to work**.

### One-to-one calls (aka Single)

You can try Giggle.js one-to-one calls using the page: `./examples/single_client.html`

Open it on 2 separate tabs or 2 different computers, and login with 2 different XMPP accounts (Jappix.com accounts only for this online demo). Then, paste the full JID of the other connected account in the call field.

The call should then be initiated, and after a while the video will come.

### Multiparty calls (aka Muji)

You can try Giggle.js multiparty calls using the page: `./examples/muji_client.html`

Open it on 3+ separate tabs or 3+ different computers, and login with 3+ different XMPP accounts (Jappix.com accounts only for this online demo). Then, join the same conference room (you can enter any room even if it doesn't exist, just be sure to use muc.jappix.com as a MUC server - eg: happy-muji-test@muc.jappix.com).

The multiparty conference call should take some time to initiate.

**Note: depending on how powerful your CPU is and how much bandwidth your network can allocated, performances may degrade with many participants.**

## Notes About Security

Jingle could have been implemented in a dirty-way there, so that "It just works". But it's **NOT** the case. Giggle.js has been thought as secure on its basis.

We wanted to avoid such possible security exploits: _imagine a friend of you wants to disturb your call, and send you a session-terminate with a wrong SID. In a quick-and-dirty implementation, the request would have been dropped because the call is already ongoing, but the client-side handlers would still have been fired, thus modifying the UI._

Giggle.js simply **DOES NOT** fire the custom event handlers that you may have defined, so that you don't have to check yourself that each incoming packet is safe, thus to ensure your client implementation of Jingle is rock-solid (safe: session is authorized **AND** stanza sender is authorized **AND** the Jingle session flow is respected).

Multiparty call can be password-protected, in a completely transparent fashion. Giggle.js can handle the password generation process, and then transmit it to other potential call participants through the Muji call invites you send to them.

## Usage

The API documentation is available in: `./doc/`. **You must first build Giggle to view it, please refer to the upper section about building Giggle.**

We assume your XMPP Web client is using a supported XMPP library and has an active connection, stored in the *GIGGLE_CONNECTION* global object.

```javascript
/*
 * Giggle.js Implementation Example
 */

// 1. Ensure you are loading the following libraries:
//    > jQuery
//    > JSJaC (or any other supported XMPP library)
//    > Giggle.js

// 2. Define configuration generators
var single_config_generator = function() {
    // Return single configuration object
    // Refer to single_client.js source for more
    return {};
};

var muji_config_generator = function() {
    // Return muji configuration object
    // Refer to muji_client.js source for more
    return {};
};

// 3. Initiate the listener
// Note: here using 'JSJaCConsoleLogger' from JSJaC as a console wrapper
Giggle.listen({
    plug: GIGGLE_PLUG_JSJAC,
    connection: GIGGLE_CONNECTION,
    debug: (new JSJaCConsoleLogger(4)),

    // Receive a one-to-one (Single) call
    single_initiate: function(stanza) {
        var config = single_config_generator();

        // Configuration values
        config.to          = stanza.from() || null;
        config.local_view  = $('#video_local')[0];
        config.remote_view = $('#video_remote')[0];

        // Handle call request
        session = Giggle.session(GIGGLE_SESSION_SINGLE, config);
        session.handle(stanza);
    },

    // Receive a multiparty (Muji) call
    muji_invite: function(stanza, args) {
        // Note: auto-accepting call there
        //       you should ask the user before accepting like this

        var config = muji_config_generator();

        // Session values
        config.to           = args.jid;
        config.media        = (args.media == GIGGLE_MEDIA_VIDEO) ? GIGGLE_MEDIA_VIDEO : GIGGLE_MEDIA_AUDIO;
        config.local_view   = $('#video_local')[0];

        if(args.password) {
            config.password = args.password;
        }

        // Handle conference invite
        session = Giggle.session(GIGGLE_SESSION_MUJI, config);
        session.join();
    }
});

// 4. Configure our disco#info handler to reply with Giggle.js features
var handle_disco_info = function(iq) {
    // Do the parse work there...

    // Get my client feature map
    var client_map = [ /* Features here... */ ];

    // Get the Giggle.js feature map
    var Giggle_map = Giggle.disco();

    // Concatenate feature maps
    var final_map = client_map.concat(Giggle_map);

    // Ensure uniqueness of feature map values there...
    // Send back your response stanza there...
};

```

### One-to-one calls

The one-to-one demo client source code can be found at the following URLs:

* **HTML**: https://github.com/valeriansaliou/giggle/blob/master/examples/single_client.html
* **CSS**: https://github.com/valeriansaliou/giggle/blob/master/examples/styles/single_client.css
* **JavaScript**: https://github.com/valeriansaliou/giggle/blob/master/examples/scripts/single_client.js

Here's a sum up on how to use the Giggle/Single API:

```javascript
/*
 * Giggle.Single.js API Example
 */

// Single call launcher
var launch_single_call = function(to, target) {
    var config = single_config_generator();

    // Session values
    config.to           = to;
    config.media        = (target == 'call_audio') ? GIGGLE_MEDIA_AUDIO : GIGGLE_MEDIA_VIDEO;
    config.video_source = (target == 'call_screen') ? GIGGLE_VIDEO_SOURCE_SCREEN : GIGGLE_VIDEO_SOURCE_CAMERA;
    config.local_view   = $('#video_local')[0];
    config.remote_view  = $('#video_remote')[0];

    // Send call request
    session = Giggle.session(GIGGLE_SESSION_SINGLE, config);
    session.initiate();
};

// Attach form submit event
$('form#launch_single_call').submit(function() {
    var to = $(this).find('input[name="to"]').val();
    var target = ($(this).find('input[name="target"]').val() || 'call_video');

    if(to && target) {
        // Launch call...
        var session = launch_single_call(to, target);

        // Send session info (client ringing & more)...
        session.info(GIGGLE_SESSION_INFO_RINGING);

        // Mute/Unmute audio in call...
        session.mute(GIGGLE_MEDIA_AUDIO);
        session.umute(GIGGLE_MEDIA_AUDIO);

        // Terminate call...
        session.terminate(GIGGLE_REASON_SUCCESS);
    }

    return false;
});

```

### Multiparty calls

The multiparty demo client source code can be found at the following URLs:

* **HTML**: https://github.com/valeriansaliou/giggle/blob/master/examples/muji_client.html
* **CSS**: https://github.com/valeriansaliou/giggle/blob/master/examples/styles/muji_client.css
* **JavaScript**: https://github.com/valeriansaliou/giggle/blob/master/examples/scripts/muji_client.js

Here's a sum up on how to use the Giggle/Muji API:

```javascript
/*
 * Giggle.Muji.js API Example
 */

// Muji call launcher
var launch_muji_call = function(room, target) {
    var config = muji_config_generator();

    // Session values
    config.to           = room;
    config.media        = (target == 'call_audio') ? GIGGLE_MEDIA_AUDIO : GIGGLE_MEDIA_VIDEO;
    config.video_source = GIGGLE_VIDEO_SOURCE_CAMERA;
    config.local_view   = $('#video_local')[0];

    // Join conference room
    session = Giggle.session(GIGGLE_SESSION_MUJI, config);
    session.join();

    return session;
};

// Attach form submit event
$('form#launch_muji_call').submit(function() {
    var room = $(this).find('input[name="room"]').val();
    var target = ($(this).find('input[name="target"]').val() || 'call_video');

    if(room && target) {
        // Launch conference...
        var session = launch_muji_call(room, target);

        // Send a message to conference...
        session.send_message('Hello World!');

        // Mute/Unmute audio in conference...
        session.mute(GIGGLE_MEDIA_AUDIO);
        session.umute(GIGGLE_MEDIA_AUDIO);

        // Invite people to conference...
        session.invite([
            'valerian@jappix.com/Phone',
            'julien@jappix.com/Tablet'
        ]);

        // Leave conference...
        session.leave();
    }

    return false;
});

```

## Ah, One More Thing...

**We worked hard on that lib, we would appreciate your feedback and bug reports.**
Plus, if you are using it in your project, we would be glad if you let [@valeriansaliou](https://valeriansaliou.name/) know.

**If you have any enhancement idea or any bug to report:** https://github.com/valeriansaliou/giggle/issues


CHANGELOG.md
Giggle.js Changelog
===================

Here's the log of what has changed over the Giggle.js releases.


v0.8.3 (Development)
--------------------

 * **In a nutshell**
    * (...)


v0.8.2 (Feb 28, 2016)
---------------------

 * **In a nutshell**
    * **Maintenance release**  @valeriansaliou


v0.8.1 (Mar 12, 2015)
---------------------

 * **In a nutshell**
    * **Fix Chrome SSRC errors** that were occuring frequently  @valeriansaliou


v0.8.0 (Mar 11, 2015)
---------------------

 * **In a nutshell**
    * **JSJaCJingle.js is now Giggle.js**              @valeriansaliou
    * **Pluggability module** to other XMPP libraries  @valeriansaliou


v0.7.6 (Oct 13, 2014)
---------------------

 * **In a nutshell**
    * **[XEP-0353](http://xmpp.org/extensions/xep-0353.html)** (full support)  @valeriansaliou
    * **WebRTC SDP messages parsing fixed** for Chrome 38+                     @valeriansaliou


v0.7.5 (Oct 3, 2014)
--------------------

 * **In a nutshell**
    * **CI links updated**  @valeriansaliou


v0.7.4 (Aug 7, 2014)
--------------------

 * **In a nutshell**
    * **NodeJS dependencies updated** (fixes broken installer)  @valeriansaliou


v0.7.3 (Jul 23, 2014)
---------------------

 * **In a nutshell**
    * **Participant stream custom handlers** added to GiggleMuji  @valeriansaliou


v0.7.2 (Jul 23, 2014)
---------------------

 * **In a nutshell**
    * **Stream custom handlers** added to GiggleSingle  @valeriansaliou


v0.7.1 (Jul 19, 2014)
---------------------

 * **In a nutshell**
    * **Quick build task** added to Gruntfile  @valeriansaliou


v0.7.0 (Jun 24, 2014)
---------------------

 * **In a nutshell**
    * **Library API changed** to support both Single (one-to-one) + Muji (one-to-many) calls     @valeriansaliou
    * **Refactored library structure** to be more maintenable (using Ring.js + GruntJS + Bower)  @valeriansaliou
    * **Code documented** using JSDoc                                                            @valeriansaliou

 * **XEP added**
    * **[XEP-0177](http://xmpp.org/extensions/xep-0177.html)** (full support)     @valeriansaliou
    * **[XEP-0249](http://xmpp.org/extensions/xep-0249.html)** (full support)     @valeriansaliou
    * **[XEP-0272](http://xmpp.org/extensions/xep-0272.html)** (full support)     @valeriansaliou
    * **[XEP-0278](http://xmpp.org/extensions/xep-0278.html)** (partial support)  @valeriansaliou
    * **[XEP-0338](http://xmpp.org/extensions/xep-0338.html)** (full support)     @valeriansaliou
    * **[XEP-0339](http://xmpp.org/extensions/xep-0339.html)** (full support)     @valeriansaliou


v0.6.0 (Dec 26, 2013)
---------------------

 * **In a nutshell**
    * **Firefox compatibility**                @valeriansaliou
    * **Safer stanza custom handler** helpers  @valeriansaliou
    * **More precise SDP debugging**           @valeriansaliou
    * **Tests for CI** added                   @valeriansaliou
    * **XEPs support file**                    @valeriansaliou


v0.5.0 (Jul 15, 2013)
---------------------

 * **In a nutshell**
    * **Fix video in Chrome 31+** loaded callback (doubling, tripling, and so forth) bug fixed  @valeriansaliou
    * **XEP support** file listing implemented protocols and specifications                     @valeriansaliou
    * **Fix wrong SDP ID/Label** which was breaking some clients                                @valeriansaliou
    * **Cross-browser compatibility** between Chrome and Firefox                                @valeriansaliou
    * **Firefox compatibility**


v0.4.0 (Jul 13, 2013)
---------------------

 * **In a nutshell**
    * Better **code debugging** for issue reporting                                                  @valeriansaliou
    * A lot of **bugfixes**                                                                          @valeriansaliou
    * **transport-info** support (add transports after descriptions are sent)                        @valeriansaliou
    * **Audio-only** calls support                                                                   @valeriansaliou
    * **Configurable video** (resolution, framerate, bandwidth)                                      @valeriansaliou
    * **Video resolution** is sent to other peer in the Jingle stanza                                @valeriansaliou
    * More **robust content naming** system (if _name_ != _media_)                                   @valeriansaliou
    * **Screen sharing** support (still experimental, requires Chrome v26+ and proper flag enabled)  @valeriansaliou


v0.3.0 (Jul 6, 2013)
--------------------

 * **In a nutshell**
    * Ability to **attach more video views** as the call is going on using _Giggle.register_view()_  @valeriansaliou

 * **XEP added**
    **[XEP-0262](http://xmpp.org/extensions/xep-0262.html)** (full support)  @valeriansaliou


v0.2.0 (Jul 2, 2013)
--------------------

 * **In a nutshell**
    * Complex **code split** in separate functions           @valeriansaliou
    * **Router fixes** (some packets were dropped)           @valeriansaliou
    * **Full integration debug**, improvements on that side  @valeriansaliou


v0.1.0 (Jun 27, 2013)
---------------------

 * **In a nutshell**
    * **Initial release**                         @valeriansaliou
    * Ability to **videochat** with another peer  @valeriansaliou
    * Working **test client**                     @valeriansaliou

 * **XEP added**
    * **[XEP-0166](http://xmpp.org/extensions/xep-0166.html)** (partial support)  @valeriansaliou
    * **[XEP-0167](http://xmpp.org/extensions/xep-0167.html)** (partial support)  @valeriansaliou
    * **[XEP-0176](http://xmpp.org/extensions/xep-0176.html)** (full support)     @valeriansaliou
    * **[XEP-0293](http://xmpp.org/extensions/xep-0293.html)** (full support)     @valeriansaliou
    * **[XEP-0294](http://xmpp.org/extensions/xep-0294.html)** (full support)     @valeriansaliou

src/giggle.main.js
/**
 * @fileoverview Giggle library - Common components
 *
 * @url https://github.com/valeriansaliou/giggle
 * @author Valérian Saliou https://valeriansaliou.name/
 *
 * @copyright 2015, Valérian Saliou
 * @license Mozilla Public License v2.0 (MPL v2.0)
 */


/** @module giggle/main */
/** @exports Giggle */


/**
 * Library main class.
 * @instance
 * @requires   nicolas-van/ring.js
 * @requires   giggle/plug
 * @requires   giggle/init
 * @requires   giggle/single
 * @requires   giggle/muji
 * @see        {@link http://ringjs.neoname.eu/|Ring.js}
 */
var Giggle = new (ring.create(
  /** @lends Giggle.prototype */
  {
    /**
     * Constructor
     */
    constructor: function(args) {
      /**
       * @constant
       * @member {GiggleUtils}
       * @readonly
       * @default
       * @public
       */
      this.utils = new GiggleUtils({
        debug: GiggleStorage.get_debug()
      });

      /**
       * @constant
       * @member {__GigglePlug}
       * @readonly
       * @default
       * @public
       */
      this.plug = {};

      /**
       * @constant
       * @member {GiggleBroadcast}
       * @readonly
       * @default
       * @public
       */
      this.broadcast = {};

      /**
       * @constant
       * @member {GiggleInit}
       * @readonly
       * @default
       * @public
       */
      this.init = {};
    },

    /**
     * Starts a new Jingle session
     * @public
     * @param {String} type
     * @param {Object} [args]
     * @returns {GiggleSingle|GiggleMuji} Giggle session instance
     */
    session: function(type, args) {
      var jingle;

      try {
        switch(type) {
          case GIGGLE_SESSION_SINGLE:
            jingle = new GiggleSingle(args);
            break;

          case GIGGLE_SESSION_MUJI:
            jingle = new GiggleMuji(args);
            break;

          default:
            throw ('Unknown session type: ' + type);
        }
      } catch(e) {
        GiggleStorage.get_debug().log('[giggle:main] session > ' + e, 1);
      } finally {
        return jingle;
      }
    },

    /**
     * Listens for Jingle events
     * @public
     * @param     {Object}           [args]
     * @property  {String}           [args.plug]             - The XMPP library to be plugged to.
     * @property  {Object}           [args.connection]       - The connection to be attached to.
     * @property  {Function}         [args.single_initiate]  - The Jingle session initiate request custom handler.
     * @property  {Function}         [args.single_propose]   - The Jingle session propose request custom handler.
     * @property  {Function}         [args.single_retract]   - The Jingle session retract request custom handler.
     * @property  {Function}         [args.single_accept]    - The Jingle session accept request custom handler.
     * @property  {Function}         [args.single_reject]    - The Jingle session reject request custom handler.
     * @property  {Function}         [args.single_proceed]   - The Jingle session proceed request custom handler.
     * @property  {Function}         [args.muji_invite]      - The Muji session invite message custom handler.
     * @property  {Console}          [args.debug]            - A reference to a debugger implementing the Console interface.
     * @property  {Boolean}          [args.extdisco]         - Whether or not to discover external services as per XEP-0215.
     * @property  {Boolean}          [args.relaynodes]       - Whether or not to discover relay nodes as per XEP-0278.
     * @property  {Boolean}          [args.fallback]         - Whether or not to request STUN/TURN from a fallback URL.
     * @see {@link https://github.com/valeriansaliou/giggle/blob/master/examples/fallback.json|Fallback JSON Sample} - Fallback URL format.
     */
    listen: function(args) {
      try {
        // Apply arguments
        if(args && args.plug)
          GiggleStorage.set_plug(args.plug);
        if(args && args.connection)
          GiggleStorage.set_connection(args.connection);
        if(args && args.single_initiate)
          GiggleStorage.set_single_initiate(args.single_initiate);
        if(args && args.single_propose)
          GiggleStorage.set_single_propose(args.single_propose);
        if(args && args.single_retract)
          GiggleStorage.set_single_retract(args.single_retract);
        if(args && args.single_accept)
          GiggleStorage.set_single_accept(args.single_accept);
        if(args && args.single_reject)
          GiggleStorage.set_single_reject(args.single_reject);
        if(args && args.single_proceed)
          GiggleStorage.set_single_proceed(args.single_proceed);
        if(args && args.muji_invite)
          GiggleStorage.set_muji_invite(args.muji_invite);
        if(args && args.debug)
          GiggleStorage.set_debug(args.debug);

        // Instanciate plug
        switch(GiggleStorage.get_plug()) {
          case GIGGLE_PLUG_JSJAC:
            this.plug = (new GigglePlugJSJaC()); break;

          default:
            GiggleStorage.get_debug().log('[giggle:main] listen > Unknown plug wrapper!', 1);
            return false;
        }

        // Instanciate broadcast
        this.broadcast = new GiggleBroadcast({
          debug : GiggleStorage.get_debug(),
          plug  : this.plug
        });

        // Instanciate initializer
        this.init = new GiggleInit({
          plug: this.plug
        });

        // Incoming IQs handler
        var cur_type, route_map = {};
        route_map[GIGGLE_STANZA_IQ]        = this._route_iq;
        route_map[GIGGLE_STANZA_MESSAGE]   = this._route_message;
        route_map[GIGGLE_STANZA_PRESENCE]  = this._route_presence;

        for(cur_type in route_map) {
          if(route_map.hasOwnProperty(cur_type)){
            this.plug.register(
              cur_type,
              route_map[cur_type].bind(this)
            );
          }
        }

        GiggleStorage.get_debug().log('[giggle:main] listen > Listening.', 2);

        // Discover available network services
        if(!args || args.extdisco !== false)
          this.init._extdisco();
        if(!args || args.relaynodes !== false)
          this.init._relaynodes();
        if(args.fallback && typeof args.fallback === 'string')
          this.init._fallback(args.fallback);
      } catch(e) {
        GiggleStorage.get_debug().log('[giggle:main] listen > ' + e, 1);
      }
    },

    /**
     * Maps the Jingle disco features
     * @public
     * @returns {Array} Feature namespaces
     */
    disco: function() {
      // Check for listen status
      var has_muji = (typeof GiggleStorage.get_muji_invite_raw() == 'function' && true);
      var has_jingle = ((has_muji || (typeof GiggleStorage.get_single_initiate_raw() == 'function')) && true);

      if(GIGGLE_AVAILABLE && has_jingle) {
        if(has_muji) {
          return MAP_DISCO_JINGLE.concat(MAP_DISCO_MUJI);
        } else {
          return MAP_DISCO_JINGLE;
        }
      }

      return [];
    },

    /**
     * Routes Jingle IQ stanzas
     * @private
     * @param {Object} stanza
     */
    _route_iq: function(stanza) {
      try {
        var from = stanza.from();

        if(from) {
          var jid_obj    = new this.utils.jid(this.utils, from);
          var from_bare  = (jid_obj.node() + '@' + jid_obj.domain());

          // Single or Muji?
          var is_muji    = (this._read(GIGGLE_SESSION_MUJI, from_bare) !== null);
          var is_single  = !is_muji;

          var action         = null;
          var sid            = null;
          var session_route  = null;

          // Route the incoming stanza
          var jingle = stanza.select_element_uniq('jingle', NS_JINGLE);

          if(jingle) {
            sid     = jingle.attribute('sid');
            action  = jingle.attribute('action');
          } else {
            var stanza_id = stanza.id();

            if(stanza_id) {
              var is_jingle = stanza_id.indexOf(GIGGLE_STANZA_ID_PRE + '_') !== -1;

              if(is_jingle) {
                var stanza_id_split = stanza_id.split('_');
                sid = stanza_id_split[1];
              }
            }
          }

          // WebRTC not available ATM?
          if(jingle && !GIGGLE_AVAILABLE) {
            GiggleStorage.get_debug().log('[giggle:main] _route_iq > Dropped Jingle packet (WebRTC not available).', 0);

            (new GiggleSingle({ to: from }))._send_error(stanza, XMPP_ERROR_SERVICE_UNAVAILABLE);
          } else if(is_muji) {
            var username, participant;

            username       = jid_obj.resource();
            session_route  = this._read(GIGGLE_SESSION_MUJI, from_bare);
            participant    = session_route.get_participants(username);

            // Muji: new session? Or registered one?
            if(participant && participant.session  &&
              (participant.session instanceof GiggleSingle)) {
              // Route to Single session
              var session_route_single = this._read(
                GIGGLE_SESSION_SINGLE,
                participant.session.get_sid()
              );

              if(session_route_single !== null) {
                GiggleStorage.get_debug().log('[giggle:main] _route_iq > [' + username + '] > Routed to Muji participant session (sid: ' + sid + ').', 2);

                session_route_single.handle(stanza);
              } else if(stanza.type() == GIGGLE_IQ_TYPE_SET && from) {
                GiggleStorage.get_debug().log('[giggle:main] _route_iq > Unknown Muji participant session route (sid: ' + sid + ').', 0);

                (new GiggleSingle({ to: from }))._send_error(stanza, GIGGLE_ERROR_UNKNOWN_SESSION);
              }
            } else if(sid) {
              if(action == GIGGLE_ACTION_SESSION_INITIATE) {
                GiggleStorage.get_debug().log('[giggle:main] _route_iq > [' + username + '] > New Muji participant session (sid: ' + sid + ').', 2);

                session_route._create_participant_session(username).handle(stanza);
              } else if(stanza.type() == GIGGLE_IQ_TYPE_SET && from) {
                GiggleStorage.get_debug().log('[giggle:main] _route_iq > Unknown Muji participant session (sid: ' + sid + ').', 0);

                (new GiggleSingle({ to: from }))._send_error(stanza, GIGGLE_ERROR_UNKNOWN_SESSION);
              }
            }
          } else if(is_single) {
            // Single: new session? Or registered one?
            session_route = this._read(GIGGLE_SESSION_SINGLE, sid);

            if(action == GIGGLE_ACTION_SESSION_INITIATE && session_route === null) {
              GiggleStorage.get_debug().log('[giggle:main] _route_iq > New Jingle session (sid: ' + sid + ').', 2);

              GiggleStorage.get_single_initiate()(stanza);
            } else if(sid) {
              if(session_route !== null) {
                GiggleStorage.get_debug().log('[giggle:main] _route_iq > Routed to Jingle session (sid: ' + sid + ').', 2);

                session_route.handle(stanza);
              } else if(stanza.type() == GIGGLE_IQ_TYPE_SET && from) {
                GiggleStorage.get_debug().log('[giggle:main] _route_iq > Unknown Jingle session (sid: ' + sid + ').', 0);

                (new GiggleSingle({ to: from }))._send_error(stanza, GIGGLE_ERROR_UNKNOWN_SESSION);
              }
            }
          } else {
            GiggleStorage.get_debug().log('[giggle:main] _route_iq > No route to session, not Jingle nor Muji (sid: ' + sid + ').', 0);
          }
        }
      } catch(e) {
        GiggleStorage.get_debug().log('[giggle:main] _route_iq > ' + e, 1);
      }
    },

    /**
     * Routes Jingle message stanzas
     * @private
     * @param {Object} stanza
     */
    _route_message: function(stanza) {
      try {
        var from = stanza.from();

        if(from) {
          var jid = new this.utils.jid(this.utils, from);

          // Broadcast message?
          var is_handled_broadcast = this.broadcast.handle(stanza);

          if(is_handled_broadcast === true) {
            // XEP-0353: Jingle Message Initiation
            // Nothing to do there.
          } else {
            // Muji?
            var room = (jid.node() + '@' + jid.domain());

            var session_route = this._read(GIGGLE_SESSION_MUJI, room);

            var x_conference  = stanza.select_element_uniq('x', NS_JABBER_CONFERENCE);
            var x_invite      = stanza.select_element_uniq('x', NS_MUJI_INVITE);

            var is_invite     = (x_conference && x_invite && true);

            if(is_invite === true) {
              if(session_route === null) {
                GiggleStorage.get_debug().log('[giggle:main] _route_message > Muji invite received (room: ' + room + ').', 2);

                // Read invite data
                var err = 0;
                var args = {
                  from     : (from                                   || err++),
                  jid      : (x_conference.attribute('jid')          || err++),
                  password : (x_conference.attribute('password')     || null),
                  reason   : (x_conference.attribute('reason')       || null),
                  media    : (x_invite.attribute('media')            || err++)
                };

                if(err === 0) {
                  GiggleStorage.get_muji_invite()(stanza, args);
                } else {
                  GiggleStorage.get_debug().log('[giggle:main] _route_message > Dropped invite because incomplete (room: ' + room + ').', 0);
                }
              } else {
                GiggleStorage.get_debug().log('[giggle:main] _route_message > Dropped invite because Muji already joined (room: ' + room + ').', 0);
              }
            } else {
              if(session_route !== null) {
                GiggleStorage.get_debug().log('[giggle:main] _route_message > Routed to Jingle session (room: ' + room + ').', 2);

                session_route.handle_message(stanza);
              }
            }
          }
        }
      } catch(e) {
        GiggleStorage.get_debug().log('[giggle:main] _route_message > ' + e, 1);
      }
    },

    /**
     * Routes Jingle presence stanzas
     * @private
     * @param {Object} stanza
     */
    _route_presence: function(stanza) {
      try {
        // Muji?
        var from = stanza.from();

        if(from) {
          var jid   = new this.utils.jid(this.utils, from);
          var room  = jid.bare();

          var session_route = this._read(GIGGLE_SESSION_MUJI, room);

          if(session_route !== null) {
            GiggleStorage.get_debug().log('[giggle:main] _route_presence > Routed to Jingle session (room: ' + room + ').', 2);

            session_route.handle_presence(stanza);
          }
        }
      } catch(e) {
        GiggleStorage.get_debug().log('[giggle:main] _route_presence > ' + e, 1);
      }
    },

    /**
     * Adds a new Jingle session
     * @private
     * @param {String} type
     * @param {String} sid
     * @param {Object} obj
     */
    _add: function(type, sid, obj) {
      GiggleStorage.get_sessions()[type][sid] = obj;
    },

    /**
     * Reads a new Jingle session
     * @private
     * @param {String} type
     * @param {String} sid
     * @returns {Object} Session
     */
    _read: function(type, sid) {
      return (sid in GiggleStorage.get_sessions()[type]) ? GiggleStorage.get_sessions()[type][sid] : null;
    },

    /**
     * Removes a new Jingle session
     * @private
     * @param {String} type
     * @param {String} sid
     */
    _remove: function(type, sid) {
      delete GiggleStorage.get_sessions()[type][sid];
    },

    /**
     * Defer given task/execute deferred tasks
     * @private
     * @param {(Function|Boolean)} arg
     */
    _defer: function(arg) {
      try {
        if(typeof arg == 'function') {
          // Deferring?
          if(GiggleStorage.get_defer().deferred) {
            (GiggleStorage.get_defer().fn).push(arg);

            GiggleStorage.get_debug().log('[giggle:main] defer > Registered a function to be executed once ready.', 2);
          }

          return GiggleStorage.get_defer().deferred;
        } else if(!arg || typeof arg == 'boolean') {
          GiggleStorage.get_defer().deferred = (arg === true);

          if(GiggleStorage.get_defer().deferred === false) {
            // Execute deferred tasks?
            if((--GiggleStorage.get_defer().count) <= 0) {
              GiggleStorage.get_defer().count = 0;

              GiggleStorage.get_debug().log('[giggle:main] defer > Executing ' + GiggleStorage.get_defer().fn.length + ' deferred functions...', 2);

              while(GiggleStorage.get_defer().fn.length)
                ((GiggleStorage.get_defer().fn).shift())();

              GiggleStorage.get_debug().log('[giggle:main] defer > Done executing deferred functions.', 2);
            }
          } else {
            ++GiggleStorage.get_defer().count;
          }
        }
      } catch(e) {
        GiggleStorage.get_debug().log('[giggle:main] defer > ' + e, 1);
      }
    },
  }
))();
